numpy/core/src/multiarray/convert_datatype.c

```diff
@@ -15,6 +15,7 @@
 #include "numpy/npy_math.h"
 
 #include "array_coercion.h"
+#include "can_cast_table.h"
```

Comment:
Is the table used here and in numpy/core/src/multiarray/legacy_dtype_implementation.c ?

Reply:
Yes, and my new scalar stuff uses it there as well (which could be up for discussion, but right now I think it makes sense).  I think it should be possible to remove it from `legacy_dtype_implementation.c` once the weak-scalar stuff goes in (or is finalized, it may be needed for warnings).

I am wondering if we should move this to `common/...`, but not sure.

On a separate note, I now think it would make sense to create a `.h.src` file with a host of inline functions: `npy_cast_@from@_to_@to@`.  It seems likely that we will have a copy of casting logic around for fast scalar math – and for now for the `dtype->casts` slot.  And it probably makes the casting code easier to read anyway.

Reply:
So merge this now and then do the refactoring later or do you want to do it now?

Reply:
Oh, I am happy to go ahead for now.  It doesn't seem terrible and I am not quite sure where to best put it/name it.  But if you have a preference, will do that.